[Agent Builder]  Observability Agent

sorenlouv opened the pull request:
This adds the plugin `ObservabilityAgent` which registers an agent for handling observability use cases for Agent Builder.
It is intentionally kept very simple to act as a foundation we can build upon.

Closes: https://github.com/elastic/obs-ai-assistant-team/issues/356

### Feature Flag
To enable the Observability Agent, add the below config in `kibana.dev.yml`
```
feature_flags.overrides:
  observabilityAgent.enabled: true
```

#### With the feature flag `ON`
<img width="1256" height="407" alt="image" src="https://github.com/user-attachments/assets/432ea8ba-4ebb-4ba5-8375-7bbccacb2721" />

#### With the feature flag `OFF`
<img width="1242" height="365" alt="image" src="https://github.com/user-attachments/assets/20da8d84-9765-4894-a916-a93510f5faa5" />

Commit:
[Agent Builder] Adds Observability Agent plugin

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-cloud-from-pr --> Click to trigger **kibana-deploy-cloud-from-pr** for this PR!

viduni94 committed:
Merge branch 'main' into agent-builder-obs-agent

Commit:
Merge branch 'main' into agent-builder-obs-agent

Commit:
Cleanup utils

Commit:
Extract `getIndexPatterns`

Commit:
Use datemath instead of ISO timestamps

Commit:
Add knowledge base tool

Commit:
Only return public entries (for now)

Commit:
Add space-awareness

Commit:
Add `get_data_sources` tool

Rename tools (remove prefix)

kibanamachine commented:
<!-- kbn-message-context:project-deploy-job -->
PR Project deployment started at: https://buildkite.com/elastic/kibana-deploy-project-from-pr/builds/687

kibanamachine commented:
<!-- kbn-message-context:project-deploy-job -->
Project deployed, see credentials at: https://buildkite.com/elastic/kibana-deploy-project-from-pr/builds/687

Commit:
Merge branch 'main' into agent-builder-obs-agent

Commit:
Merge branch 'main' of github.com:elastic/kibana into agent-builder-obs-agent

Commit:
Add feature flag to enable agent

```yaml
uiSettings:
  overrides:
    'observability:registerObservabilityAgent': true
```

Commit:
Change prompt

viduni94 commented on src/platform/packages/shared/kbn-management/settings/setting_ids/index.ts:
```diff
@@ -128,6 +128,8 @@ export const OBSERVABILITY_AI_ASSISTANT_SIMULATED_FUNCTION_CALLING =
   'observability:aiAssistantSimulatedFunctionCalling';
 export const OBSERVABILITY_AI_ASSISTANT_SEARCH_CONNECTOR_INDEX_PATTERN =
   'observability:aiAssistantSearchConnectorIndexPattern';
+export const OBSERVABILITY_REGISTER_OBSERVABILITY_AGENT_ID =
```
Is there a reason to call this `AGENT_ID`?

If not, what do you think about:
```suggestion
export const OBSERVABILITY_REGISTER_OBSERVABILITY_AGENT =
```

viduni94 commented on x-pack/solutions/observability/plugins/observability_agent/server/register_observability_agent.ts:
```diff
@@ -0,0 +1,104 @@
+    description: OBSERVABILITY_AGENT_DESCRIPTION,
+    avatar_icon: 'logoObservability',
+    configuration: {
+      instructions: 'You are a observability specialist agent',
```
```suggestion
      instructions: 'You are an observability specialist agent',
```

viduni94 commented on x-pack/solutions/observability/plugins/observability_agent/server/tools/get_data_sources.ts:
```diff
@@ -0,0 +1,85 @@
+            {
+              type: ToolResultType.error,
+              data: {
+                message: `Failed to retrieve data sources: ${error.message}`,
```
```suggestion
                message: `Failed to retrieve observability data sources: ${error.message}`,
```

viduni94 commented on x-pack/solutions/observability/plugins/observability_agent/server/tools/get_services.ts:
```diff
@@ -0,0 +1,184 @@
+
+export const OBSERVABILITY_GET_SERVICES_TOOL_ID = 'observability.get_services';
+
+const createServiceInventorySchema = (indexPatterns: string[]) =>
```
For the `get_data_sources` tool, the schema was named as `getDataSourcesSchema`.
Can we rename this to `getServicesSchema` to keep it consistent?

viduni94 commented on x-pack/solutions/observability/plugins/observability_agent/server/tools/recall_knowledge_base.ts:
```diff
@@ -0,0 +1,277 @@
+          logger,
+        });
+
+        if (entries.length === 0) {
```
In future, are we planning to surface a knowledge base installation path to the user, if they haven't already?

viduni94 commented on x-pack/solutions/observability/plugins/observability_agent/server/utils/dsl_filters.ts:
```diff
@@ -0,0 +1,55 @@
+  return [{ term: { 'service.environment': environment } }];
+}
+
+// { term: { 'metricset.name': 'service_transaction' } },
```
Commented line

viduni94 commented on x-pack/solutions/observability/plugins/observability_agent/server/utils/get_logs_indices.ts:
```diff
@@ -0,0 +1,42 @@
+  ObservabilityAgentPluginStartDependencies,
+} from '../types';
+
+export const DEFAULT_LOG_INDEX_PATTERNS = ['logs-*'];
```
This doesn't seem to be used anywhere

viduni94 committed:
Merge branch 'main' into agent-builder-obs-agent

sorenlouv committed:
Update x-pack/solutions/observability/plugins/observability_agent/server/register_observability_agent.ts

Co-authored-by: Viduni Wickramarachchi <viduni.ushanka@gmail.com>

sorenlouv committed:
Update x-pack/solutions/observability/plugins/observability_agent/server/register_observability_agent.ts

Co-authored-by: Viduni Wickramarachchi <viduni.ushanka@gmail.com>

sorenlouv committed:
Update x-pack/solutions/observability/plugins/observability_agent/server/register_observability_agent.ts

Co-authored-by: Viduni Wickramarachchi <viduni.ushanka@gmail.com>

Commit:
Merge branch 'main' of github.com:elastic/kibana into agent-builder-obs-agent

Commit:
Add support for private entries

Commit:
Address feedback

Commit:
Remove `get_services` tool

Commit:
Rephrase “recall” to “search”

elasticmachine commented:
Pinging @elastic/obs-ux-management-team (Team:obs-ux-management)

Commit:
Update codeowners

Commit:
Fix tool names

sorenlouv commented on x-pack/solutions/observability/plugins/observability/server/ui_settings.ts:
```diff
@@ -435,6 +436,21 @@ export const uiSettings: Record<string, UiSettingsParams<boolean | number | stri
+    }),
+    description: i18n.translate('xpack.observability.registerObservabilityAgent.description', {
+      defaultMessage:
+        'Enable the Observability Agent in Agent Builder. Restarting Kibana is required.',
```
@viduni94 If you add a feature flag we may as well remove the ui setting

neptunian replied:
> Wouldn't we still need the UI setting so that users can turn the obs-agent "on" to opt-in?

No, I believe there is going to be a single shared advanced setting that all the solutions can use.  In classic mode they would all change together.   So I don't think we need our own UI setting.

viduni94 replied:
Thanks @neptunian 
The UI setting is removed - https://github.com/elastic/kibana/pull/237630/commits/6ad1c57cf0f6e70067adca6d0d9040df6a3acacf

sorenlouv committed:
Update x-pack/solutions/observability/plugins/observability_agent/server/tools/get_data_sources/get_data_sources.ts

Co-authored-by: Viduni Wickramarachchi <viduni.ushanka@gmail.com>

sorenlouv committed:
Update x-pack/solutions/observability/plugins/observability_agent/server/tools/get_data_sources/get_data_sources.ts

Co-authored-by: Viduni Wickramarachchi <viduni.ushanka@gmail.com>

sorenlouv committed:
Update x-pack/solutions/observability/plugins/observability_agent/server/tools/get_data_sources/get_data_sources.ts

Co-authored-by: Viduni Wickramarachchi <viduni.ushanka@gmail.com>

sorenlouv committed:
Update x-pack/solutions/observability/plugins/observability_agent/server/tools/search_knowledge_base/search_knowledge_base.ts

Co-authored-by: Viduni Wickramarachchi <viduni.ushanka@gmail.com>

kibanamachine committed:
Changes from node scripts/lint_ts_projects --fix

kibanamachine committed:
Changes from node scripts/build_plugin_list_docs

kibanamachine committed:
TO FIX: Run node 'scripts/yarn_deduplicate && yarn kbn bootstrap' locally, or add an exception to src/dev/yarn_deduplicate/index.ts and then commit the changes and push to your branch

viduni94 committed:
Fix lint error by handling the promise returned by getIsObservabilityAgentEnabled(core)

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/359120)
* Commit: 9fd3db0fe6a97f3b2580be0b0eb05d178ab5f53d
* [Cloud Deployment](https://kibana-pr-237630-402862.kb.us-west2.gcp.elastic-cloud.com)

### History
* :broken_heart: [Build #359085](https://buildkite.com/elastic/kibana-pull-request/builds/359085) failed e9691027343992bdbe09e16423e7ac5bcde7fca8

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"running","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/359120","number":359120,"commit":"9fd3db0fe6a97f3b2580be0b0eb05d178ab5f53d"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/359085","number":359085,"commit":"e9691027343992bdbe09e16423e7ac5bcde7fca8"}],"number":359120}
buildkite-pr-comment-->

viduni94 committed:
Add agent feature flag

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/359147)
* Commit: eb611e4ffdd3eb286e26f1c22e8a1b1b78d17fea
* [Cloud Deployment](https://kibana-pr-237630-402862.kb.us-west2.gcp.elastic-cloud.com)

### History
* :broken_heart: [Build #359120](https://buildkite.com/elastic/kibana-pull-request/builds/359120) failed 9fd3db0fe6a97f3b2580be0b0eb05d178ab5f53d
* :broken_heart: [Build #359085](https://buildkite.com/elastic/kibana-pull-request/builds/359085) failed e9691027343992bdbe09e16423e7ac5bcde7fca8

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"running","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/359147","number":359147,"commit":"eb611e4ffdd3eb286e26f1c22e8a1b1b78d17fea"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/359120","number":359120,"commit":"9fd3db0fe6a97f3b2580be0b0eb05d178ab5f53d"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/359085","number":359085,"commit":"e9691027343992bdbe09e16423e7ac5bcde7fca8"}],"number":359147}
buildkite-pr-comment-->

viduni94 committed:
Remove UI setting and add README

kibanamachine committed:
Changes from node scripts/lint_ts_projects --fix

viduni94 committed:
Merge branch 'main' into agent-builder-obs-agent

kibanamachine committed:
Changes from node scripts/build_plugin_list_docs

viduni94 committed:
Update README description

kibanamachine committed:
Changes from node scripts/build_plugin_list_docs

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/359201)
* Commit: 0a9b1f39b06fc9e1bf682169ca9d9f93793ac929
* [Cloud Deployment](https://kibana-pr-237630-402862.kb.us-west2.gcp.elastic-cloud.com)

### History
* :green_heart: [Build #359186](https://buildkite.com/elastic/kibana-pull-request/builds/359186) succeeded 5517ded642c6cd9524c6c1afa4916dac9b141b30
* :broken_heart: [Build #359147](https://buildkite.com/elastic/kibana-pull-request/builds/359147) failed eb611e4ffdd3eb286e26f1c22e8a1b1b78d17fea
* :broken_heart: [Build #359120](https://buildkite.com/elastic/kibana-pull-request/builds/359120) failed 9fd3db0fe6a97f3b2580be0b0eb05d178ab5f53d
* :broken_heart: [Build #359085](https://buildkite.com/elastic/kibana-pull-request/builds/359085) failed e9691027343992bdbe09e16423e7ac5bcde7fca8

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"running","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/359201","number":359201,"commit":"0a9b1f39b06fc9e1bf682169ca9d9f93793ac929"},{"buildStatus":{"state":"running","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/359186","number":359186,"commit":"5517ded642c6cd9524c6c1afa4916dac9b141b30"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/359147","number":359147,"commit":"eb611e4ffdd3eb286e26f1c22e8a1b1b78d17fea"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/359120","number":359120,"commit":"9fd3db0fe6a97f3b2580be0b0eb05d178ab5f53d"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/359085","number":359085,"commit":"e9691027343992bdbe09e16423e7ac5bcde7fca8"}],"number":359201}
buildkite-pr-comment-->

viduni94 commented on x-pack/solutions/observability/plugins/observability_agent/server/tools/get_data_sources/get_data_sources.ts:
```diff
@@ -0,0 +1,85 @@
+        const { apmIndices, logIndexPatterns, metricIndexPatterns, alertsIndexPattern } =
+          await getObservabilityDataSources({ core, plugins, logger });
+
+        return {
```
Yes, that sounds good to me. We can add it as a follow up if you prefer @sorenlouv 
Created an issue https://github.com/elastic/obs-ai-assistant-team/issues/400

Commit:
Remove query rewriting

kibanamachine committed:
Changes from node scripts/lint_ts_projects --fix

Commit:
Document feature flag in readme

Commit:
Remove instructions to add link to  knowledge base

Commit:
Merge branch 'main' into agent-builder-obs-agent

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/359434)
* Commit: b727ce98f87a30f7b378ae7d4c95a30db557edf5
* [Cloud Deployment](https://kibana-pr-237630-402862.kb.us-west2.gcp.elastic-cloud.com)

### History
* :green_heart: [Build #359270](https://buildkite.com/elastic/kibana-pull-request/builds/359270) succeeded 77554fa8c980cd4cae580ea8ac8d41a162a7aa0b
* :green_heart: [Build #359201](https://buildkite.com/elastic/kibana-pull-request/builds/359201) succeeded 0a9b1f39b06fc9e1bf682169ca9d9f93793ac929
* :green_heart: [Build #359186](https://buildkite.com/elastic/kibana-pull-request/builds/359186) succeeded 5517ded642c6cd9524c6c1afa4916dac9b141b30
* :broken_heart: [Build #359147](https://buildkite.com/elastic/kibana-pull-request/builds/359147) failed eb611e4ffdd3eb286e26f1c22e8a1b1b78d17fea
* :broken_heart: [Build #359120](https://buildkite.com/elastic/kibana-pull-request/builds/359120) failed 9fd3db0fe6a97f3b2580be0b0eb05d178ab5f53d

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"running","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/359434","number":359434,"commit":"b727ce98f87a30f7b378ae7d4c95a30db557edf5"},{"buildStatus":{"state":"running","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/359270","number":359270,"commit":"77554fa8c980cd4cae580ea8ac8d41a162a7aa0b"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/359201","number":359201,"commit":"0a9b1f39b06fc9e1bf682169ca9d9f93793ac929"},{"buildStatus":{"state":"running","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/359186","number":359186,"commit":"5517ded642c6cd9524c6c1afa4916dac9b141b30"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/359147","number":359147,"commit":"eb611e4ffdd3eb286e26f1c22e8a1b1b78d17fea"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/359120","number":359120,"commit":"9fd3db0fe6a97f3b2580be0b0eb05d178ab5f53d"}],"number":359434}
buildkite-pr-comment-->

viduni94 committed:
Remove redundant const

arturoliduena reviewed (approved):
LGTM, good call moving it under a feature flag instead of using UI settings.

viduni94 committed:
Merge branch 'main' into agent-builder-obs-agent

pgayvallet commented on x-pack/solutions/observability/plugins/observability_agent/server/tools/register_tools.ts:
```diff
@@ -0,0 +1,56 @@
+  const observabilityTools: StaticToolRegistration<any>[] = await Promise.all([
+    createGetDataSourcesTool({ core, plugins, logger }),
+    createSearchKnowledgeBaseTool({ core, logger }),
+  ]);
```
NIT: looking at the implem of both tools, it looks like the factory methods don't really need to be async?

sorenlouv committed:
Address feedback

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/360017)
* Commit: 45bb7a5b1bcde15254d5a04cacf6ad362ee7b233
* [Cloud Deployment](https://kibana-pr-237630-6a0c2d.kb.us-west2.gcp.elastic-cloud.com)

### History
* :yellow_heart: [Build #359861](https://buildkite.com/elastic/kibana-pull-request/builds/359861) was flaky 9bc89146740597494e3322cf75eec05c57fd0f82
* :green_heart: [Build #359496](https://buildkite.com/elastic/kibana-pull-request/builds/359496) succeeded bdf5a00f09115bd236e5bf126438f22103effb00
* :yellow_heart: [Build #359434](https://buildkite.com/elastic/kibana-pull-request/builds/359434) was flaky b727ce98f87a30f7b378ae7d4c95a30db557edf5
* :green_heart: [Build #359270](https://buildkite.com/elastic/kibana-pull-request/builds/359270) succeeded 77554fa8c980cd4cae580ea8ac8d41a162a7aa0b
* :green_heart: [Build #359201](https://buildkite.com/elastic/kibana-pull-request/builds/359201) succeeded 0a9b1f39b06fc9e1bf682169ca9d9f93793ac929

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"running","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/360017","number":360017,"commit":"45bb7a5b1bcde15254d5a04cacf6ad362ee7b233"},{"buildStatus":{"state":"running","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/359861","number":359861,"commit":"9bc89146740597494e3322cf75eec05c57fd0f82"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/359496","number":359496,"commit":"bdf5a00f09115bd236e5bf126438f22103effb00"},{"buildStatus":{"state":"running","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/359434","number":359434,"commit":"b727ce98f87a30f7b378ae7d4c95a30db557edf5"},{"buildStatus":{"state":"running","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/359270","number":359270,"commit":"77554fa8c980cd4cae580ea8ac8d41a162a7aa0b"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/359201","number":359201,"commit":"0a9b1f39b06fc9e1bf682169ca9d9f93793ac929"}],"number":360017}
buildkite-pr-comment-->

neptunian committed:
Apply suggestion from @arturoliduena

Co-authored-by: Arturo Lidueña <arturo.liduena@elastic.co>

neptunian commented on x-pack/solutions/observability/plugins/observability_agent/server/utils/get_observability_data_sources.ts:
```diff
@@ -0,0 +1,43 @@
+}> {
+  const apmIndexPatterns = await getApmIndices({ core, plugins, logger });
+  const logIndexPatterns = await getLogsIndices({ core, logger });
+  const metricIndexPatterns = ['metrics-*'];
```
We should add the metrics equivalent in a followup.

sorenlouv replied:
I see now that `x-pack/solutions/observability/plugins/metrics_data_access` exists We should use that similar to `apmDataAccess` and `logsDataAccess`

neptunian commented on x-pack/solutions/observability/plugins/observability_agent/server/tools/register_tools.ts:
```diff
@@ -0,0 +1,56 @@
+
+const PLATFORM_TOOL_IDS = [
+  platformCoreTools.search,
+  platformCoreTools.listIndices,
```
I think having both are useful. The agent will call `get_data_sources` when mentioning observability data like logs, metrics, etc but will call `list_indices` when mentioning a specific index pattern.
